### PR TITLE
Improve error message when saving products using invalid user

### LINKED ISSUE #247: `saveProduct` provides poor error message for invalid users
**Describe the bug**

The error message produced when trying to save a product using an invalid user is poor.

The problem is that the SDK always assumes that the API respone is valid JSON, which is not true. See the "Status Codes" part described here:

https://wiki.openfoodfacts.org/API/Write 

Problematic part in the SDK:

```dart
var status = Status.fromJson(json.decode(response.body));
return status;
```
This code is located here https://github.com/openfoodfacts/openfoodfacts-dart/blob/master/lib/openfoodfacts.dart#L108-L109

In consequence the JSON conversion fails and an exception is thrown that does not indicate username/password problems.

**To Reproduce**
Issue can be produced using the following unit test, which currently fails:

```dart
    String _generateRandomString(int len) {
      var r = Random();
      return String.fromCharCodes(
          List.generate(len, (index) => r.nextInt(33) + 89));
    }

    test('Invalid user saves product', () async {
      Product product = Product(
        barcode: '7340011364184',
      );
      Status status = await OpenFoodAPIClient.saveProduct(
        User(
          userId: 'invaliduser',
          password: _generateRandomString(16),
        ),
        product,
      );
     // Exception is thrown, so assertions have been left out
    });
  });

```

**Expected behavior**
Expected the returned `Status` to contain information that indicates username/password problems.

**package information:**
Openfoodfacts package version: 1.3.4

### PR DESCRIPTION
**Pull request by peterwvj**
Fixes #247 and gets the tests to pass.

### CODE REVIEW

**Review by monsieurtanuki — COMMENTED** [review]
Hi @peterwvj!

That's a start, but I presume that such a typical error (wrong username/password) should be dealt with more gracefully by the backend.
Or if not by the backend, by us.

What about an additional `factory Status.fromJsonString(String jsonString)` that would
* try to decode the jsonString and call `Status.fromJson` (like today)
* catch an exception and "grep" `'Incorrect user name or password.'` to display a specific "wrong user/password" status error
* displays a generic error when the string is not found

For those who did not try this error, with a wrong login/password we get an HTML page like this one:
<details>
<summary>HTML</summary>

```html
<!-- start templates/web/common/site_layout.tt.html -->

<!doctype html>
<html class="no-js" lang="en" data-serverdomain="openfoodfacts.net" dir="ltr">
<head>
    <meta charset="utf-8">
    <title>Error</title>
    
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
	<meta property="fb:app_id" content="219331381518041">
    <meta property="og:type" content="food">
    <meta property="og:title" content="">
    <meta property="og:url" content="https://world.openfoodfacts.net">
    
    <meta property="og:image" content="https://world.openfoodfacts.org/images/misc/openfoodfacts-logo-en-356.png">
    <meta property="og:description" content="">
    <link rel="manifest" href="/cgi/manifest.pl">
<link rel="mask-icon" href="/images/favicon/safari-pinned-tab.svg" color="#5bbad5">
<link rel="shortcut icon" href="/images/favicon/favicon.ico">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="msapplication-TileImage" content="/images/favicon/mstile-144x144.png">
<meta name="msapplication-config" content="/images/favicon/browserconfig.xml">
<meta name="_globalsign-domain-verification" content="2ku73dDL0bAPTj_s1aylm6vxvrBZFK59SfbH_RdUya" />
<meta name="apple-itunes-app" content="app-id=588797948">
<meta name="flattr:id" content="dw637l">

    <link rel="canonical" href="https://world.openfoodfacts.net">
    <link rel="stylesheet" href="https://static.openfoodfacts.net/css/dist/app-ltr.css?v=1630928770">
    <link rel="stylesheet" href="https://static.openfoodfacts.net/css/dist/jqueryui/themes/base/jquery-ui.css">
    <link rel="stylesheet" href="https://static.openfoodfacts.net/css/dist/select2.min.css">
    <link rel="search" href="https://world.openfoodfacts.net/cgi/opensearch.pl" type="application/opensearchdescription+xml" title="Open Food Facts">
    
    <style media="all">
        
        
    </style>
    <!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-31851927-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-31851927-1');
</script>


</head>
<body class="error_page">
<nav class="top-bar" data-topbar id="top-bar">
	<ul class="title-area">
		<li class="name">
			<h2><a href="/" style="font-size:1rem;">Open Food Facts</a></h2>
		</li>
		<li class="toggle-topbar menu-icon">
			<a href="#"><span>Menu</span></a>
		</li>
	</ul>
	<section class="top-bar-section">
    <label for="select_country" style="display:none">Country</label>
    <ul class="country_language_selection">
			<li class="has-form has-dropdown" id="select_country_li">
				<select id="select_country" style="width:100%" data-placeholder="Country">
					<option></option>
				</select>
			</li>
      
    </ul>
    <ul class="search_and_links">
      <li class="show-for-large-up">
        <form action="/cgi/search.pl">
          <div class="row collapse">
            <div class="small-8 columns">
              <input type="text" placeholder="Search for a product" name="search_terms" value="">
              <input name="search_simple" value="1" type="hidden">
              <input name="action" value="process" type="hidden">
            </div>
            <div class="small-4 columns">
              <button type="submit" title="Search"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0 0 16 9.5 6.5 6.5 0 1 0 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"/><path d="M0 0h24v24H0z" fill="none"/></svg></button>
            </div>
          </div>
        </form>
      </li>
      <li class="show-for-large-only"><a href="/cgi/search.pl" title="Advanced search"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"/><path d="M0 0h24v24H0z" fill="none"/></svg></a></li>
      <li class="show-for-xlarge-up"><a href="/cgi/search.pl"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"/><path d="M0 0h24v24H0z" fill="none"/></svg> Advanced search</span></a></li>
      <li class="show-for-large-only"><a href="/cgi/search.pl?graph=1" title="Graphs and maps"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M5 9.2h3V19H5zM10.6 5h2.8v14h-2.8zm5.6 8H19v6h-2.8z"/><path fill="none" d="M0 0h24v24H0z"/></svg></a></li>
      <li class="show-for-xlarge-up"><a href="/cgi/search.pl?graph=1"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M5 9.2h3V19H5zM10.6 5h2.8v14h-2.8zm5.6 8H19v6h-2.8z"/><path fill="none" d="M0 0h24v24H0z"/></svg> Graphs and maps</span></a></li>
      
        <li class="show-for-large-up divider"></li>
        <li><a href="/discover">Discover</a></li>
        <li><a href="/contribute">Contribute</a></li>
        <li class="show-for-large"><a href="/open-food-facts-mobile-app" title="Get the app" class="button success"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M16 1H8C6.34 1 5 2.34 5 4v16c0 1.66 1.34 3 3 3h8c1.66 0 3-1.34 3-3V4c0-1.66-1.34-3-3-3zm-2 20h-4v-1h4v1zm3.25-3H6.75V4h10.5v14z"/><path d="M0 0h24v24H0z" fill="none"/></svg></a></li>
        <li class="show-for-xlarge-up"><a href="/open-food-facts-mobile-app" class="button success"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M16 1H8C6.34 1 5 2.34 5 4v16c0 1.66 1.34 3 3 3h8c1.66 0 3-1.34 3-3V4c0-1.66-1.34-3-3-3zm-2 20h-4v-1h4v1zm3.25-3H6.75V4h10.5v14z"/><path d="M0 0h24v24H0z" fill="none"/></svg> Get the app</a></li>
      
    </ul>
	</section>
</nav>



<nav class="tab-bar show-for-small-only">
	<div class="left-small">
		<a href="#idOfLeftMenu" role="button" aria-controls="idOfLeftMenu" aria-expanded="false" class="left-off-canvas-toggle button postfix anonymous">
		<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M3 5v14a2 2 0 0 0 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2H5a2 2 0 0 0-2 2zm12 4c0 1.66-1.34 3-3 3s-3-1.34-3-3 1.34-3 3-3 3 1.34 3 3zm-9 8c0-2 4-3.1 6-3.1s6 1.1 6 3.1v1H6v-1z"/><path d="M0 0h24v24H0z" fill="none"/></svg>
		</a>
	</div>
	<div class="middle tab-bar-section">
		<form action="/cgi/search.pl">
			<div class="row collapse">
				<div class="small-8 columns">
					<input type="text" placeholder="Search for a product" name="search_terms">
					<input name="search_simple" value="1" type="hidden">
					<input name="action" value="process" type="hidden">
				</div>
				<div class="small-2 columns">
					<button type="submit" class="button postfix"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0 0 16 9.5 6.5 6.5 0 1 0 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"/><path d="M0 0h24v24H0z" fill="none"/></svg></button>
				</div>
				<div class="small-2 columns">
					<a href="/cgi/search.pl" title="Advanced search"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0 0 16 9.5 6.5 6.5 0 1 0 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"/><path d="M0 0h24v24H0z" fill="none"/></svg> <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"/><path d="M0 0h24v24H0z" fill="none"/></svg></a>
				</div>
			</div>
		</form>
	</div>
</nav>

<div class="off-canvas-wrap" data-offcanvas>
	<div class="inner-wrap">
		<aside class="left-off-canvas-menu">
			<div id="aside_column">
				<!-- start templates/web/common/includes/display_blocks.tt.html -->



    <div class="block">
        
            <h3 class="block_title">Sign in</h3>
        
        <div class="block_content">
            <!-- start templates/web/common/includes/display_login_register.tt.html -->

<p>Sign-in to add or edit products.</p>

<form method="post" action="/cgi/session.pl">
	<div class="row">
		<div class="small-12 columns">
			<label>Username or e-mail address:
				<input type="text" name="user_id" autocomplete="username" required>
			</label>
		</div>
		<div class="small-12 columns">
			<label>Password
				<input type="password" name="password" autocomplete="current-password" required>
			</label>
		</div>
		<div class="small-12 columns">
			<label>
				<input type="checkbox" name="remember_me" checked>
				Remember me
			</label>
		</div>
	</div>
	<input type="submit" name=".submit" value="Sign in" class="button small">
</form>


<p>Not registered yet?
    <a href="/cgi/user.pl">Create your account. </a><br>
    <a href="/cgi/reset_password.pl">Forgotten password? </a>
</p>

<!-- end templates/web/common/includes/display_login_register.tt.html -->
        </div>
    </div>

    



    <div class="block">
        
            <h3 class="block_title">Add a product</h3>
        
        <div class="block_content">
            <!-- start templates/web/common/includes/display_product_search_or_add.tt.html -->


  <div id="imgsearchdiv_block_aside">

<a href="#" class="button small expand" id="imgsearchbutton_block_aside"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><circle cx="12" cy="12" r="3.2"/><path d="M9 2 7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z"/><path d="M0 0h24v24H0z" fill="none"/></svg> Picture with barcode
<input type="file" accept="image/*" class="img_input" name="imgupload_search" id="imgupload_search_block_aside" style="position: absolute;
    right:0;
    bottom:0;
    top:0;
    cursor:pointer;
    opacity:0;
    font-size:40px;"/>
</a>
</div>

<div id="progressbar_block_aside" class="progress" style="display:none">
  <span id="progressmeter_block_aside" class="meter" style="width:0%"></span>
</div>

<div id="imgsearchmsg_block_aside" data-alert class="alert-box info" style="display:none">
  Sending image
  <a href="#" class="close">&times;</a>
</div>

<div id="imgsearcherror_block_aside" data-alert class="alert-box alert" style="display:none">
  Upload error
  <a href="#" class="close">&times;</a>
</div>




<form method="post" action="/cgi/product.pl" enctype="multipart/form-data">

    <div class="row collapse">
        <div class="small-9 columns">
          <input type="text" name="code" placeholder="or Barcode">
        </div>
        <div class="small-3 columns">
           <input type="submit" value="Add" class="button postfix">
        </div>
    </div>

	<input type="submit" value="Product without barcode" class="button tiny">

</form>

<!-- end templates/web/common/includes/display_product_search_or_add.tt.html -->
        </div>
    </div>

    



    <div class="block">
        
            <h3 class="block_title">Donate to support our work</h3>
        
        <div class="block_content">
            <a href="https://world.openfoodfacts.org/donate-to-open-food-facts"><img src="https://static.openfoodfacts.org/images/svg/donate-icon.svg" alt="Donate to Open Food Facts" /></a><p>Open Food Facts is made by a non-profit association, independent from the industry. It is made for all, by all, and it is funded by all. You can support our work by <a href="https://world.openfoodfacts.org/donate-to-open-food-facts">donating to Open Food Facts</a> and also by <a href="https://www.lilo.org/fr/open-food-facts/?utm_source=open-food-facts">using the Lilo search engine</a>.<br/><b>Thank you!</b></p>
        </div>
    </div>

    


<!-- end templates/web/common/includes/display_blocks.tt.html -->
			</div>
		</aside>
		<a class="exit-off-canvas"></a>
		
		<!-- full width banner on mobile -->
		
		
		
		<!-- main row - comment used to remove left column and center content on some pages -->
		<div class="row full-width" style="max-width: 100% !important;" data-equalizer>
			<div class="xxlarge-2 xlarge-2 large-3 medium-4 columns hide-for-small" style="background-color:#fafafa;padding-top:1rem;" data-equalizer-watch>
				<div class="sidebar">
					<div style="text-align:center">
						<a href="/"><img id="logo" src="/images/misc/openfoodfacts-logo-en-178x150.png" srcset="/images/misc/openfoodfacts-logo-en-356x300.png 2x" width="178" height="150" alt="Open Food Facts logo" style="margin-bottom:0.5rem"></a>
					</div>
					Open Food Facts gathers information and data on food products from around the world.
					<form action="/cgi/search.pl" class="hide-for-large-up">
						<div class="row collapse">
							<div class="small-9 columns">
								<input type="text" placeholder="Search for a product" name="search_terms">
								<input name="search_simple" value="1" type="hidden">
								<input name="action" value="process" type="hidden">
							</div>
							<div class="small-2 columns">
								<button type="submit" class="button postfix"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0 0 16 9.5 6.5 6.5 0 1 0 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"/><path d="M0 0h24v24H0z" fill="none"/></svg></button>
							</div>
							<div class="small-1 columns">
								<label class="right inline">
									<a href="/cgi/search.pl" title="Advanced search"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"/><path d="M0 0h24v24H0z" fill="none"/></svg></a>
								</label>
							</div>
						</div>
					</form>
					<!-- start templates/web/common/includes/display_blocks.tt.html -->



    <div class="block">
        
            <h3 class="block_title">Sign in</h3>
        
        <div class="block_content">
            <!-- start templates/web/common/includes/display_login_register.tt.html -->

<p>Sign-in to add or edit products.</p>

<form method="post" action="/cgi/session.pl">
	<div class="row">
		<div class="small-12 columns">
			<label>Username or e-mail address:
				<input type="text" name="user_id" autocomplete="username" required>
			</label>
		</div>
		<div class="small-12 columns">
			<label>Password
				<input type="password" name="password" autocomplete="current-password" required>
			</label>
		</div>
		<div class="small-12 columns">
			<label>
				<input type="checkbox" name="remember_me" checked>
				Remember me
			</label>
		</div>
	</div>
	<input type="submit" name=".submit" value="Sign in" class="button small">
</form>


<p>Not registered yet?
    <a href="/cgi/user.pl">Create your account. </a><br>
    <a href="/cgi/reset_password.pl">Forgotten password? </a>
</p>

<!-- end templates/web/common/includes/display_login_register.tt.html -->
        </div>
    </div>

    



    <div class="block">
        
            <h3 class="block_title">Add a product</h3>
        
        <div class="block_content">
            <!-- start templates/web/common/includes/display_product_search_or_add.tt.html -->


  <div id="imgsearchdiv_block_side">

<a href="#" class="button small expand" id="imgsearchbutton_block_side"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" class="icon" aria-hidden="true" focusable="false"><circle cx="12" cy="12" r="3.2"/><path d="M9 2 7.17 4H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2h-3.17L15 2H9zm3 15c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5z"/><path d="M0 0h24v24H0z" fill="none"/></svg> Picture with barcode
<input type="file" accept="image/*" class="img_input" name="imgupload_search" id="imgupload_search_block_side" style="position: absolute;
    right:0;
    bottom:0;
    top:0;
    cursor:pointer;
    opacity:0;
    font-size:40px;"/>
</a>
</div>

<div id="progressbar_block_side" class="progress" style="display:none">
  <span id="progressmeter_block_side" class="meter" style="width:0%"></span>
</div>

<div id="imgsearchmsg_block_side" data-alert class="alert-box info" style="display:none">
  Sending image
  <a href="#" class="close">&times;</a>
</div>

<div id="imgsearcherror_block_side" data-alert class="alert-box alert" style="display:none">
  Upload error
  <a href="#" class="close">&times;</a>
</div>




<form method="post" action="/cgi/product.pl" enctype="multipart/form-data">

    <div class="row collapse">
        <div class="small-9 columns">
          <input type="text" name="code" placeholder="or Barcode">
        </div>
        <div class="small-3 columns">
           <input type="submit" value="Add" class="button postfix">
        </div>
    </div>

	<input type="submit" value="Product without barcode" class="button tiny">

</form>

<!-- end templates/web/common/includes/display_product_search_or_add.tt.html -->
        </div>
    </div>

    



    <div class="block">
        
            <h3 class="block_title">Donate to support our work</h3>
        
        <div class="block_content">
            <a href="https://world.openfoodfacts.org/donate-to-open-food-facts"><img src="https://static.openfoodfacts.org/images/svg/donate-icon.svg" alt="Donate to Open Food Facts" /></a><p>Open Food Facts is made by a non-profit association, independent from the industry. It is made for all, by all, and it is funded by all. You can support our work by <a href="https://world.openfoodfacts.org/donate-to-open-food-facts">donating to Open Food Facts</a> and also by <a href="https://www.lilo.org/fr/open-food-facts/?utm_source=open-food-facts">using the Lilo search engine</a>.<br/><b>Thank you!</b></p>
        </div>
    </div>

    


<!-- end templates/web/common/includes/display_blocks.tt.html -->
				</div>
			</div>
			<div id="main_column" class="xxlarge-10 xlarge-10 large-9 medium-8 columns" style="padding-top:1rem" data-equalizer-watch>
				
				<!-- main column content - comment used to remove left column and center content on some pages -->
				
				
<!-- start templates/web/common/includes/donate_banner.tt.html -->



<div id="donate_banner" class="row">
	<div class="small-12 large-12 xlarge-8 xxlarge-7 columns">
		<div id="image_banner" style="margin-bottom:1rem;" class="panel callout">
		
          
			<h2></h2>
			<p><br>
			
			</p>
			<div class="row">
				<div class="small-12 large-4 columns">
			<a href="https://world.openfoodfacts.org/donate-to-open-food-facts?utm_source=off&utm_medium=web&utm_campaign=donate-2020&utm_term=en-text-button" class="button round" style="margin-bottom:0">
				
			</a>

		
				</div>
				<div class="small-12 large-8 columns">
					<input id="hide_image_banner" type="checkbox" style="margin-top:1.5rem;margin-bottom:0;">
					<label for="hide_image_banner">
						<span id="hide_image_banner_hide">I have already donated or I'm not interested. Hide the banner.</span>
					</label>
				</div>
			</div>
		</div>
	</div>
</div>


<!-- end templates/web/common/includes/donate_banner.tt.html -->

				
				<h1>Error</h1>

				<p>Incorrect user name or password. <a href="/cgi/reset_password.pl">Forgotten password?</a></p>
				
			</div>
        </div>
	</div>
</div>

<footer>
	<div class="small-12 medium-6 large-3 columns off">
		<div class="title">Open Food Facts</div>
		<p>A collaborative, free and open database of food products from around the world.</p>
		<ul>
			<li><a href="/legal">Legal</a></li>
			<li><a href="/terms-of-use">Terms of use</a></li>
			<li><a href="/data">Data, API and SDKs</a></li>
			<li><a href="https://world.openfoodfacts.org/donate-to-open-food-facts">Donate to Open Food Facts</a></li>
			<li><a href="http://world.pro.openfoodfacts.org/">Producers</a></li>
		</ul>
	</div>
	<div class="small-12 medium-6 large-3 columns app">
		<div class="title">Install the app</div>
		<a href="https://apps.apple.com/app/open-food-facts/id588797948"><img src="/images/misc/appstore/black/appstore_US.svg" alt="Download on the App Store" width="120" height="40" loading="lazy"></a>
		<a href="https://play.google.com/store/apps/details?id=org.openfoodfacts.scanner&hl=en"><img src="https://static.openfoodfacts.org/images/misc/google-play-badge-svg-master/img/en_get.svg" alt="Get It On Google Play" width="102" height="40" loading="lazy"></a>
		<a href="https://www.microsoft.com/en-us/p/openfoodfacts/9nblggh0dkqr"><img src="/images/misc/microsoft/English.svg" alt="Get it from Microsoft" width="109" height="40" loading="lazy"></a>
		<a href="https://world.openfoodfacts.org/files/off.apk"><img src="https://static.openfoodfacts.org/images/misc/android-apk.svg" alt="Android APK" loading="lazy"></a>
	</div>
	<div class="small-12 medium-6 large-3 columns project">
		<div class="title">Discover the project</div>
		<ul>
			<li><a href="/who-we-are">Who we are</a></li>
			<li><a href="https://world.openfoodfacts.org/open-food-facts-vision-mission-values-and-programs">Vision, Mission, Values and Programs</a></li>
			<li><a href="/faq">Frequently asked questions</a></li>
			<li><a href="https://en.blog.openfoodfacts.org">Open Food Facts blog</a></li>
			<li><a href="/press">Press</a></li>
			<li><a href="https://wiki.openfoodfacts.org">Open Food Facts wiki (en)</a></li>
			<li><a href="/cgi/top_translators.pl">Translators</a></li>
			<li><a href="/partners">Partners</a></li>
			<li><a href="https://world-fr.openbeautyfacts.org">Open Beauty Facts - Cosmétiques</a></li>
		</ul>
	</div>
	<div class="small-12 medium-6 large-3 columns community">
		<div class="title">Join the community</div>
		<p>
			<a href="/code-of-conduct">Code of conduct</a>
			<br><br>
			Join us on <a href="https://slack.openfoodfacts.org">Slack</a>:
			<br>
			and the <a href="https://www.facebook.com/groups/openfoodfacts/">Facebook group for contributors</a>
			Follow us on <a href="https://twitter.com/openfoodfacts">Twitter</a>,
<a href="https://www.facebook.com/OpenFoodFacts">Facebook</a> and
<a href="https://www.instagram.com/open.food.facts/">Instagram</a>

		</p>
	</div>
</footer>

<script src="https://static.openfoodfacts.net/js/dist/modernizr.js"></script>
<script src="https://static.openfoodfacts.net/js/dist/jquery.js"></script>
<script src="https://static.openfoodfacts.net/js/dist/jquery-ui.js"></script>
<script src="https://static.openfoodfacts.net/js/dist/display.js"></script>

<script>
$(function() {

// start off canvas blocks for small screens

    $('#imgupload_search_block_side').fileupload({
		sequentialUploads: true,
        dataType: 'json',
        url: '/cgi/product.pl',
		formData : [{name: 'jqueryfileupload', value: 1}],
		resizeMaxWidth : 2000,
		resizeMaxHeight : 2000,
        done: function (e, data) {
			if (data.result.location) {
				$(location).attr('href',data.result.location);
			}
			if (data.result.error) {
				$("#imgsearcherror_block_side").html(data.result.error);
				$("#imgsearcherror_block_side").show();
			}
        },
		fail : function (e, data) {
			$("#imgsearcherror_block_side").show();
        },
		always : function (e, data) {
			$("#progressbar_block_side").hide();
			$("#imgsearchbutton_block_side").show();
			$("#imgsearchmsg_block_side").hide();
        },
		start: function (e, data) {
			$("#imgsearchbutton_block_side").hide();
			$("#imgsearcherror_block_side").hide();
			$("#imgsearchmsg_block_side").show();
			$("#progressbar_block_side").show();
			$("#progressmeter_block_side").css('width', "0%");

		},
            sent: function (e, data) {
                if (data.dataType &&
                        data.dataType.substr(0, 6) === 'iframe') {
                    // Iframe Transport does not support progress events.
                    // In lack of an indeterminate progress bar, we set
                    // the progress to 100%, showing the full animated bar:
                    $("#progressmeter_block_side").css('width', "100%");
                }
            },
            progress: function (e, data) {

                   $("#progressmeter_block_side").css('width', parseInt(data.loaded / data.total * 100, 10) + "%");
					$("#imgsearchdebug_block_side").html(data.loaded + ' / ' + data.total);

            }

    });

// end off canvas blocks for small screens

if ($.cookie('hide_image_banner_2020') == '1') {
	$('#hide_image_banner').prop('checked', true);
	$('#donate_banner').remove();
}
else {
	$('#hide_image_banner').prop('checked', false);
	$('#donate_banner').show();
	$('#hide_image_banner').change(function () {
		if ($('#hide_image_banner').prop('checked')) {
			$.cookie('hide_image_banner_2020', '1', { expires: 90, path: '/' });
			$('#donate_banner').remove();
		}
	});
}


});
</script>



<script src="https://static.openfoodfacts.net/js/dist/foundation.js"></script>
<script src="https://static.openfoodfacts.net/js/dist/jquery.cookie.js"></script>
<script src="https://static.openfoodfacts.net/js/dist/select2.min.js"></script>
<script type="text/javascript" src="/js/dist/jquery.iframe-transport.js"></script>
<script type="text/javascript" src="/js/dist/jquery.fileupload.js"></script>
<script type="text/javascript" src="/js/dist/load-image.all.min.js"></script>
<script type="text/javascript" src="/js/dist/canvas-to-blob.js"></script>

<script>
$(document).foundation({
	equalizer : {
		equalize_on_stack: true
	},
	accordion: {
		callback : function (accordion) {
			$(document).foundation('equalizer', 'reflow');
		}
	}
});

</script>
<script type="application/ld+json">
{
	"@context" : "https://schema.org",
	"@type" : "WebSite",
	"name" : "Open Food Facts",
	"url" : "https://world.openfoodfacts.net",
	"potentialAction": {
		"@type": "SearchAction",
		"target": "https://world.openfoodfacts.net/cgi/search.pl?search_terms=?{search_term_string}",
		"query-input": "required name=search_term_string"
	}
}
{
	"@context": "https://schema.org/",
	"@type": "Organization",
	"url": "https://world.openfoodfacts.net",
	"logo": "/images/misc/openfoodfacts-logo-en-178x150.png",
	"name": "Open Food Facts",
	"sameAs" : [ "https://www.facebook.com/OpenFoodFacts", "https://twitter.com/OpenFoodFacts"]
}
</script>





</body>
</html>
```

</details>

**Comment by peterwvj** [comment]
@monsieurtanuki thanks for your feedback. I updated the PR based on your suggestions. Please take a look, thanks.

**Review by monsieurtanuki — COMMENTED** [review]
Thank you @peterwvj for your code changes.

`Status.fromJson` is still used in `OpenFoodAPIClient.postInsightAnnotation`; I guess your new `Status.fromApiResponse` would make sense instead. We could almost deprecate `Status.fromJson` - I've said _almost_, I'm not 100% sure about it.

I'm a bit confused with what I read in `OpenFoodAPIClient.register`, where "they" have a similar algorithm with `status.body!.contains('...')`. In their case, they put a `400` status code, and a label in English. In your case, you put a `-1` status code, a label in English, and a `bool` method. A bit incoherent. For the moment, could you just change your status code from `-1` to `400`?

I guess `Status` could be refactored a bit further, either using your `bool isWrongUsernameOrPassword()` style, or with more elaborate status codes, or with a new `enum StatusReturnCode`. But that should be part of a different issue.

**Comment by peterwvj** [comment]
Thanks for your feedback @monsieurtanuki .

I updated the PR based on your input:
- the PR now uses `400` for status code and
- `postInsightAnnotation` now uses `fromApiRespnse`.

From what I can tell there is still a `Status.fromJson` use case that seems valid but could perhaps be moved into `fromApiReponse`. But that could be part of a new PR.

Are you fine with approving the PR now?

**Review by M123-dev — APPROVED** [review]
Thanks for the PR @peterwvj, after the comments of monsieurtanuki everything looks good to me. But I agree with you we should definitely rework the status so that this remains consistent throughout the package.

**Review by monsieurtanuki — APPROVED** [review]
That's perfect @peterwvj!
The additional refactoring can take place in a next PR.

File: lib/model/Status.dart [hunk 1]
```diff
@@ -27,5 +32,34 @@ class Status extends JsonObject {
...
+    } catch (e) {
+      return Status(
+        body: responseBody,
+        status: 400,
+        statusVerbose: _createStatusVerbose(responseBody, e),
+      );
+    }
+  }
+
+  bool isWrongUsernameOrPassword() =>
```
[the remark this answers is not in the document]
Comment (peterwvj) [reply lib/model/Status.dart]: Good catch, fixed that :)

**Pull request merged**